packages/kbn-config-schema/src/types/object_type.ts

```diff
@@ -77,7 +100,54 @@ export class ObjectType<P extends Props = any> extends Type<ObjectResultType<P>>
       .options({ stripUnknown: { objects: unknowns === 'ignore' } });
 
     super(schema, typeOptions);
-    this.props = schemaKeys;
+    this.props = props;
+    this.propSchemas = schemaKeys;
+    this.options = options;
+  }
+
+  /**
+   * Return a new `ObjectType` instance extended with given `newProps` properties.
+   * Original properties can be deleted from the copy by passing a `null` or `undefined` value for the key.
+   *
+   * @example
+   * How to add a new key to an object schema
+   * ```ts
+   * const origin = schema.object({
+   *   initial: schema.string(),
+   * });
+   *
+   * const extended = origin.extends({
+   *   added: schema.number(),
+   * });
+   * ```
+   *
+   * How to remove an existing key from an object schema
+   * ```ts
+   * const origin = schema.object({
+   *   initial: schema.string(),
+   *   toRemove: schema.number(),
+   * });
+   *
+   * const extended = origin.extends({
+   *   toRemove: undefined,
+   * });
+   * ```
+   */
+  public extends<NP extends NullableProps>(newProps: NP): ExtendedObjectType<P, NP> {
+    const extendedProps = Object.entries({
+      ...this.props,
+      ...newProps,
+    }).reduce((memo, [key, value]) => {
+      if (value !== null && value !== undefined) {
+        return {
+          ...memo,
+          [key]: value,
+        };
+      }
+      return memo;
+    }, {} as ExtendedProps<P, NP>);
+
+    return new ObjectType(extendedProps, this.options as ExtendedObjectTypeOptions<P, NP>);
```

Comment:
I need to forcecast the current options as the new type's.  The only situation where this would be unsafe is if there is a `defaultValue` or `validate` method defined at the object's level, as the original value could potentially not match the new schema, and therefor return an invalid value when validating undefined properties with the new schema.

I.E

```ts
const origin = schema.object({
  initial: schema.string(),
}, {defaultValue: {initial: 'foo'} });

const extended = origin.extends({
    added: schema.number(),
});

// validated = {initial: 'foo'} here, which does not respect `TypeOf<typeof extended>`
// and will be incorrectly typed as a valid `TypeOf<typeof extended>`
const validated = extended.validate(undefined);
```

No sure this is a blocker. One option is to not use the original object's options and use `{}` instead, but I'm not sure this is really a good solution. Another one would be to have `extends` have a **mandatory** second parameter that would be the new options. But either way, we wouldn't have a good way to duplicate the original options in the extended schema...

Reply:
It means that `extend` doesn't support `options` customization. Can we infer options correct type from `extendedProps`?
```typescript
  public extends<NP extends NullableProps>(
    newProps: NP,
    options?: ObjectTypeOptions<ExtendedProps<P, NP>>
  ): ExtendedObjectType<P, NP> {
    const extendedProps = Object.entries({
      ...this.props,
      ...newProps,
    }).reduce((memo, [key, value]) => {
      if (value !== null && value !== undefined) {
        return {
          ...memo,
          [key]: value,
        };
      }
      return memo;
    }, {} as ExtendedProps<P, NP>);

    return new ObjectType(extendedProps, {
      ...options,
      ...this.options,
    } as ObjectTypeOptions<ExtendedProps<P, NP>>);
  }
```
Then I can see errors in both cases:
```typescript
    const origin = schema.object(
      {
        initial: schema.string(),
        another: schema.maybe(schema.string()),
      },
      { defaultValue: { initial: 'foo' } }
    );

    origin.extends(
      {
        initial: schema.number(),
      },
      { defaultValue: { initial: '2' } } // <<< Types of property 'initial' are incompatible. Type 'string' is not assignable to type 'number'.
    );
    origin.extends(
      {
        initial: schema.number(),
      },
      { defaultValue: { initial: 2, another: 2 } } // <<< Types of property 'another' are incompatible. Type 'number' is not assignable to type 'string | undefined'
    );
```

Reply:
> It means that extend doesn't support options customization. Can we infer options correct type from extendedProps?

We can, and the PR is already doing it (`ExtendedObjectTypeOptions<P, NP>` in)

https://github.com/elastic/kibana/blob/148cfff67e69adcbe4eed4f9ca6bc2b836cbadb4/packages/kbn-config-schema/src/types/object_type.ts#L150

I can definitely add `options?` as a second parameter to `extends`, however if the parameter is optional, that doesn't fix the issue of my example in https://github.com/elastic/kibana/pull/68067#discussion_r434487728, as `defaultValue` from the initial schema would/could still be used by the extended schema even if it doesn't match the extended schema.

```ts
{
      ...options,
      ...this.options,
    } as ObjectTypeOptions<ExtendedProps<P, NP>>
``` 

still requires a force-cast, as the `defaultValue` and `validate` from `options` does not match the signature of the equivalent expected types in `ObjectTypeOptions<ExtendedProps<P, NP>>`

Note that in real usages, it's very uncommon to specify a defaultValue for a root config object, so it's probably a 'minor' issue, but still.

Reply:
>  that doesn't fix the issue of my example in #68067 (comment), as defaultValue from the initial schema would/could still be used by the extended schema even if it doesn't match the extended schema.

Right, but a user has an option to rewrite `defaultValue` at least.